src/core/model/reducer.ts

```diff
@@ -92,11 +76,21 @@ export const uiSchemaReducer = (
 
 export const combinedReducer = (state: EditorState, action: CombinedAction) => {
   switch (action.type) {
+    case SET_SCHEMA:
+      return linkSchemas(
+        buildSchemaTree(action.schema),
+        cloneDeep(state.uiSchema)
+      );
+    case SET_UISCHEMA:
+      return linkSchemas(
+        cleanLinkedElements(state.schema),
+        buildEditorUiSchemaTree(action.uiSchema)
+      );
     case SET_SCHEMAS:
-      return {
-        schema: buildSchemaTree(action.schema),
-        uiSchema: buildEditorUiSchemaTree(action.uiSchema),
-      };
+      return linkSchemas(
+        buildSchemaTree(action.schema),
+        buildEditorUiSchemaTree(action.uiSchema)
+      );
```

Comment:
Together with the previous comment I would replace this with
```typescript
    case SET_SCHEMA:
      return withCloneTree(state.uiSchema, state, (clonedUiSchema) => {
        return linkSchemas(buildSchemaTree(action.schema), clonedUiSchema);
      });
    case SET_UISCHEMA:
      return withCloneTree(state.schema, state, (clonedSchema) => {
        return linkSchemas(
          clonedSchema,
          buildEditorUiSchemaTree(action.uiSchema)
        );
      });
    case SET_SCHEMAS:
      return linkSchemas(
        buildSchemaTree(action.schema),
        buildEditorUiSchemaTree(action.uiSchema)
      );
```
The `linkSchemas` makes sure that the models are consistent afterwards.

If you prefer using the `clean`-approach then we could also adjust `linkSchemas` to expect completely clean schemas. Then we would need to add respective clean methods here, i.e.

```typescript
    case SET_SCHEMA:
      return withCloneTree(state.uiSchema, state, (clonedUiSchema) => {
        return linkSchemas(buildSchemaTree(action.schema), cleanUiSchemaLinks(clonedUiSchema));
      });
    case SET_UISCHEMA:
      return withCloneTree(state.schema, state, (clonedSchema) => {
        return linkSchemas(
          cleanSchemaLinks(clonedSchema),
          buildEditorUiSchemaTree(action.uiSchema)
        );
      });
    case SET_SCHEMAS:
      return linkSchemas(
        buildSchemaTree(action.schema),
        buildEditorUiSchemaTree(action.uiSchema)
      );
```

The clean methods should then not clone themselves I guess.

Reply:
I added the new method to clean the UI schema and kept the calls to the `clean` methods in the reducer, because we don't need to clean both schemas each time the `linkElements` method is called.